US-6: Destination Building Travel Plan

cristian-aldea opened the pull request:
This PR aims to add functionality to the sliding panel so that the user can start a travel plan by first bringing up the sliding panel, and then pressing the start travel button.

This completes the required functionality of #73, which includes the functionality already implemented by #74.

The liberty was taken to fix/improve the styling of the styling panel, and remove some of the hardcoded values that make the styling break sometimes.

Some of the code of the sliding panel was moved to other files to make it slimmer and easier to read.


This PR implements issue #73

cristian-aldea committed:
Fix and cleanup sliding panel styling #73

cristian-aldea committed:
Implement start travel plan button #73

r-saba commented:
Can you move the button to the top left so the user does not have to slide the panel up to click it? Have but travel button, chevron and close button on one line is what I'm recommending.
![Screenshot_20200411-161608](https://user-images.githubusercontent.com/25574985/79054137-6ee45480-7c10-11ea-8418-03850802702c.png)

r-saba commented:
As a student, I want to choose a destination building to travel to so that I can end my travel plan.

### Acceptance Criteria
* Destination building can be filled in by:
   * tapping on building polygon/marker (in omnibox focus)
   * searching for a building (in omnibox focus)
   * tapping `Set Destination` on the sliding panel
* This assumes outdoor to outdoor navigation
* Must follow [DoD](https://github.com/RGPosadas/WayFinder/wiki/Definition-of-Done)
* Must match [Mockup](https://www.figma.com/file/1XeyBVJ2XyC6AHDHYwxvdG/WayFinder-SOEN390?node-id=610%3A1)


### User Acceptance Flow
#### Omnibox Search option
* User taps on search box
* User can search for a building to set as their desired destination (in search box focus)

#### Omnibox Tap option
* User taps on search box
* User taps on desired buliding to be autofilled in the touched textbox

#### Tapping on building option
* User taps on a building
* Sliding panel pops up with a button giving the user the option to travel to that building, which starts a travel plan

r-saba commented on __tests__/building-information.test.js:
```diff
@@ -48,4 +24,23 @@ describe("BuildingInformation component", () => {
+      .onPressOut();
+
+    expect(mockOnBuildingStartTravelPlan).toHaveBeenCalledTimes(1);
+  });
```
Can you also test the parameter passed into the function is what you expect? This type of testing helped to find errors in US-5 when a function was changed.

r-saba commented on src/components/building-information/list-element.component.tsx:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @param linkItem The link item to display
+ */
+const LinkElement = ({ linkItem }: IProps) => {
```
Component name and file name are different and confusing. Please use the same words for both.

cristian-aldea replied:
Forgot to change it after a rename, good catch.

r-saba commented on __tests__/list-element.test.js:
```diff
@@ -0,0 +1,28 @@
+    expect(tree).toMatchSnapshot();
+  });
+
+  it("should check if clicking department links is functional", () => {
+    const wrapper = shallow(<LinkElement linkItem={mockLinkItem} />);
+
+    const link = wrapper.find({ testID: "linkItem.title" });
+
+    expect(link.props().children).toEqual(mockLinkItem.title);
+    link.simulate("press");
```
How exactly is this checking that clicking a link is functional? The press is done at the very end and nothing is checked after it.

cristian-aldea replied:
Sorry this is how the original test was, I managed to figure out how to mock Linking for this.

r-saba replied:
Thank you for looking into how to test links!

r-saba commented on src/components/building-information/building-information.component.tsx:
```diff
@@ -70,110 +78,56 @@ class BuildingInformation extends React.Component<IProps, IState> {
+                  style={styles.travelButton}
+                  testID="building-info-start-travel"
+                  onPressOut={() => {
+                    onBuildingStartTravelPlan(building);
```
Not a fan of this function name. How about `startBuildingTravelPlan(building)` or something similar?

TheGreatMarkus replied:
onStartBuildingTravelPlan?

cristian-aldea replied:
ok sure

r-saba replied:
Thanks!

r-saba commented on src/components/building-information/building-information.component.tsx:
```diff
@@ -70,110 +78,56 @@ class BuildingInformation extends React.Component<IProps, IState> {
+                    onBuildingStartTravelPlan(building);
+                  }}
+                >
+                  <Text style={{ color: "white" }}>Travel Here!</Text>
```
No inline styles please.

cristian-aldea replied:
Outdated.

r-saba commented on src/components/building-information/building-information.component.tsx:
```diff
@@ -211,31 +180,18 @@ const styles = StyleSheet.create({
     color: "white",
```
Please put this in a Colours file. We may want to change our app colours or prefer a slightly grey hue instead of white in the future and this will make it difficult since we'd have to go find every "white" and change it.

cristian-aldea replied:
Not really applicate this late in the project but sure

r-saba replied:
I do think good practices and standards should be followed at all stages of development. Thanks for changing even though you disagree.

cristian-aldea committed:
Adjust styling #73

r-saba commented on src/components/building-information/list-element.component.tsx:
```diff
@@ -0,0 +1,49 @@
+  },
+  linkText: {
+    fontSize: 16,
+    color: "blue",
```
Colours file please. We may have more links in the future and having to put blue or even worse having different colours used for all of them will make our app feel inconsistent.

cristian-aldea replied:
Not really applicate this late in the project but sure

r-saba replied:
I do think good practices and standards should be followed at all stages of development. Thanks for changing even though you disagree.

r-saba commented on src/components/search/autocomplete.component.tsx:
```diff
@@ -42,7 +37,7 @@ const AutoComplete = ({ searchResults, style, setLocation }: IProps) => {
           <TouchableOpacity
             testID="touchableList"
-            onPressOut={() => setLocation(item)}
+            onPress={() => setLocation(item)}
```
Did you test these changes with a Samsung phone? I remember `onPressOut` was added specifically to resolve an issue with Samsung devices.

JimLau49 replied:
Tested on a Samsung S8, functionality appears unaffected with the touchableopacity.

JimLau49 replied:
I said i was gonna throw in a comment about this on slack but forgot till now 😅

cristian-aldea replied:
I didn't, that issue occurred when using react-native's API, and it seems that gesture-handler's API is more reliable.

I'll put it back to onPressOut since it doesn't really affect me either way, but it doesn't work for me if we were to go back to react-native's TouchableOpacity.

r-saba replied:
Thanks for testing @JimLau49 . @TheGreatMarkus Agreed, gesture-handler seems to be the recommended library to use for components with user interaction.

r-saba reviewed (changes requested):
Good PR, just some minor changes and some more thorough testing I'd like to see. Can you also remove unused imports as well? I see one in `building.information.tsx`.
![image](https://user-images.githubusercontent.com/25574985/79054886-b2da5800-7c16-11ea-8d94-842046fa32db.png)

cristian-aldea committed:
Address pr feedback #73

r-saba commented on src/components/building-information/building-information.component.tsx:
```diff
@@ -68,112 +76,67 @@ class BuildingInformation extends React.Component<IProps, IState> {
+              </View>
+              <Text
+                testID="panelDisplayName"
+                style={{
+                  ...styles.buildingName,
+                  fontSize: building.displayName.length > 30 ? 18 : 24,
+                }}
```
Not a blocker but is there a reason you're not using flatten here? You've specifically requested that it be used over this method, does it not work here?

cristian-aldea replied:
true, i gotta do what I preach.

r-saba reviewed (dismissed):
LGTM! The sliding panel works well and destination can be set as per the acceptance criteria.

### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [x] All acceptance tests pass (See notes)
- [x] If ticket is a user story, there must exist an automated acceptance test for it

### Acceptance Criteria
Destination building can be filled in by:
- [x] tapping on building polygon/marker (in omnibox focus)
- [x] searching for a building (in omnibox and search focus)
- [x] tapping `Set Destination` on the sliding panel

### User Acceptance Flow
#### Search option
- [x] User taps on search box
- [x] User can search for a building to set as their desired destination (in search box focus)

#### Omnibox Tap/Search option
- [x] User taps on search box
- [x] User brings out the omnibox
- [x] User taps on desired buliding to be autofilled in the destination location textbox
- [x] User can also search for desired destination building  again in the omnibox

#### Tapping on Building option
- [x] User taps on a building
- [x] Sliding panel pops up with a button giving the user the option to travel to that building, which starts a travel plan

### Notes
LGTM! I did not run e2e tests since I don't own a mac but viewed the [gif](https://github.com/RGPosadas/WayFinder/wiki/Acceptance-Tests-GIFs#us-6)

r-saba reviewed (dismissed):
LGTM!

RGPosadas reviewed (changes requested):
### Breaking Existing Implementation for US-3
#### Description
If user is in a building, then the sliding panel containing the information of that building must pop up (implemented in #16). This currently gives an error.

#### Steps to Recreate the Error
1. Mock your location to a building
```tsx
// in map-screen.component.tsx, mock your location to Hall

  const setStartLocationBuilding = (coordinates: Location) => {
    // Attempt to find the building the user is in.
    const isInBuilding = Buildings.find((building) =>
      isPointInPolygon({ latitude: 45.497237, longitude: -73.578913 }, building.boundingBox)
    );
    if (isInBuilding) {
      onBuildingTap(isInBuilding);
      setStartLocation(buildingToMarker(isInBuilding));
    } else {
      setStartLocation(
        UtilityService.getInstance().locationToSearchResult(
          CURRENT_LOCATION_DISPLAY_TEXT,
          CURRENT_LOCATION_DISPLAY_TEXT,
          coordinates
        )
      );
    }
  };
```

2. Tap on the location button

#### Error Log
```
undefined is not an object (evaluating 'building.displayName')
* src/components/building-information/building-information.component.tsx:88:23 in BuildingInformation#render
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:13042:21 in finishClassComponent
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:12970:4 in updateClassComponent
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:307:15 in invokeGuardedCallbackImpl
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:531:36 in invokeGuardedCallback
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:20488:8 in beginWork$$1
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:19370:24 in performUnitOfWork
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:19347:39 in workLoopSync
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:18997:22 in renderRoot
* [native code]:null in renderRoot
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:18709:28 in runRootCallback
* [native code]:null in runRootCallback
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:5642:32 in runWithPriority$argument_1
- node_modules/scheduler/cjs/scheduler.development.js:643:23 in unstable_runWithPriority
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:5638:22 in flushSyncCallbackQueueImpl
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:5627:28 in flushSyncCallbackQueue
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:18556:30 in scheduleUpdateOnFiber
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:11484:17 in dispatchAction
* [native code]:null in dispatchAction
* src/screens/map-screen/map-screen.component.tsx:82:24 in onBuildingTap
* src/screens/map-screen/map-screen.component.tsx:230:4 in setStartLocationBuilding
* src/screens/map-screen/map-screen.component.tsx:211:27 in LocationService.getInstance.getCurrentLocationAsync.then$argument_0
- node_modules/promise/setimmediate/core.js:37:14 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:25 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:146:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:194:17 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:458:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:407:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:384:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:142:17 in __guard$argument_0
* [native code]:null in flushedQueue
* [native code]:null in invokeCallbackAndReturnFlushedQueue
```

cristian-aldea committed:
Fix bug caused by rebase #73

TheGreatMarkus commented:
@RGPosadas 

good catch! I didn't test the branch after the rebase. Should be fixed now.

RGPosadas commented:
> @RGPosadas
> 
> good catch! I didn't test the branch after the rebase. Should be fixed now.

Just tested, it works now

osahtout commented on src/components/building-information/link-element.component.tsx:
```diff
@@ -0,0 +1,50 @@
+import { LINK_COLOR } from "../../styles";
+
+interface IProps {
+  linkItem: LinkItem;
```
Can I ask for the name 'linkItem' to be changed to hyperlinkItem, urlLinkItem or something similar? linkItem can be interpreted as linking between two items

ritchellesap committed:
modify US-4 acceptance test, add US-6 acceptance test and GIF #73

Also modified the names for the link items. They must be unique for detox to click on them.

RGPosadas reviewed (approved):
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [x] All acceptance tests pass
- [x] If ticket is a user story, there must exist an automated acceptance test for it

### Acceptance Criteria
Destination building can be filled in by:
- [x] tapping on building polygon/marker (in omnibox focus)
- [x] searching for a building (in omnibox and search focus)
- [x] tapping `Set Destination` on the sliding panel

### User Acceptance Flow
#### Search option
- [x] User taps on search box
- [x] User can search for a building to set as their desired destination (in search box focus)

#### Omnibox Tap/Search option
- [x] User taps on search box
- [x] User brings out the omnibox
- [x] User taps on desired buliding to be autofilled in the destination location textbox
- [x] User can also search for desired destination building  again in the omnibox

#### Tapping on Building option
- [x] User taps on a building
- [x] Sliding panel pops up with a button giving the user the option to travel to that building, which starts a travel plan

### Notes
Good work, lgtm! 🚀

bit172 reviewed (approved):
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [x] All acceptance tests pass
- [x] If ticket is a user story, there must exist an automated acceptance test for it

### Acceptance Criteria
Destination building can be filled in by:
- [x] tapping on building polygon/marker (in omnibox focus)
- [x] searching for a building (in omnibox and search focus)
- [x] tapping `Set Destination` on the sliding panel

### User Acceptance Flow
#### Search option
- [x] User taps on search box
- [x] User can search for a building to set as their desired destination (in search box focus)

#### Omnibox Tap/Search option
- [x] User taps on search box
- [x] User brings out the omnibox
- [x] User taps on desired buliding to be autofilled in the destination location textbox
- [x] User can also search for desired destination building  again in the omnibox

#### Tapping on Building option
- [x] User taps on a building
- [x] Sliding panel pops up with a button giving the user the option to travel to that building, which starts a travel plan

### Notes
LGTM

leojrsilao reviewed (approved):
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [x] All acceptance tests pass
- [x] If ticket is a user story, there must exist an automated acceptance test for it

### Acceptance Criteria
Destination building can be filled in by:
- [x] tapping on building polygon/marker (in omnibox focus)
- [x] searching for a building (in omnibox and search focus)
- [x] tapping `Set Destination` on the sliding panel

### User Acceptance Flow
#### Search option
- [x] User taps on search box
- [x] User can search for a building to set as their desired destination (in search box focus)

#### Omnibox Tap/Search option
- [x] User taps on search box
- [x] User brings out the omnibox
- [x] User taps on desired buliding to be autofilled in the destination location textbox
- [x] User can also search for desired destination building  again in the omnibox

#### Tapping on Building option
- [x] User taps on a building
- [x] Sliding panel pops up with a button giving the user the option to travel to that building, which starts a travel plan

### Notes
Good work ! Everything LGTM :racing_car:  :racing_car:  :racing_car:
